x/streamer/keeper/distribute.go

```diff
@@ -3,143 +3,156 @@ package keeper
-			continue
+			return sdk.Coins{}, fmt.Errorf("add rewards to gauge %d: %w", record.GaugeId, err)
 		}
-		totalDistributedCoins = totalDistributedCoins.Add(streamDistributedCoins...)
+
+		totalAllocated = totalAllocated.Add(allocatedCoin)
 	}
 
-	return totalDistributedCoins, nil
+	return totalAllocated, nil
 }
 
-// distributeStream runs the distribution logic for a stream, and adds the sends to
-// the distrInfo struct. It also updates the stream for the distribution.
-func (k Keeper) distributeStream(ctx sdk.Context, stream types.Stream) (sdk.Coins, error) {
-	totalDistrCoins := sdk.NewCoins()
-	remainCoins := stream.Coins.Sub(stream.DistributedCoins...)
-	remainEpochs := stream.NumEpochsPaidOver - stream.FilledEpochs
-
-	for _, coin := range remainCoins {
-		epochAmt := coin.Amount.Quo(sdk.NewInt(int64(remainEpochs)))
-		if epochAmt.IsPositive() {
-			totalDistrCoins = totalDistrCoins.Add(sdk.Coin{Denom: coin.Denom, Amount: epochAmt})
+// UpdateStreamAtEpochStart updates the stream for a new epoch.
+func (k Keeper) UpdateStreamAtEpochStart(ctx sdk.Context, stream types.Stream) (types.Stream, error) {
+	// Check if stream has completed its distribution
+	if stream.FilledEpochs >= stream.NumEpochsPaidOver {
+		err := k.moveActiveStreamToFinishedStream(ctx, stream)
+		if err != nil {
+			return types.Stream{}, fmt.Errorf("move active stream to finished stream: %w", err)
 		}
+		return stream, nil
 	}
 
+	// If the stream is not finalized, update it for the next distribution
+
+	remainCoins := stream.Coins.Sub(stream.DistributedCoins...)
+	remainEpochs := stream.NumEpochsPaidOver - stream.FilledEpochs
+	epochCoins := remainCoins.QuoInt(math.NewIntFromUint64(remainEpochs))
+
 	// If the stream uses a sponsorship plan, query it and update stream distr info. The distribution
 	// might be empty and this is a valid scenario. In that case, we'll just skip at without
 	// filling the epoch.
 	if stream.Sponsored {
 		distr, err := k.sk.GetDistribution(ctx)
 		if err != nil {
-			return sdk.Coins{}, fmt.Errorf("failed to get sponsorship distribution: %w", err)
+			return types.Stream{}, fmt.Errorf("get sponsorship distribution: %w", err)
 		}
-		info := types.DistrInfoFromDistribution(distr)
 		// Update stream distr info
-		stream.DistributeTo = info
+		stream.DistributeTo = types.DistrInfoFromDistribution(distr)
 	}
 
-	totalDistrCoins, err := k.DistributeByWeights(ctx, totalDistrCoins, stream.DistributeTo)
-	if err != nil {
-		return nil, err
-	}
+	// Add coins to distribute during the next epoch
+	stream.EpochCoins = epochCoins
 
-	err = k.updateStreamPostDistribute(ctx, stream, totalDistrCoins)
-	if err != nil {
-		return nil, err
+	// Don't fill streams in which there's nothing to fill. Note that rewards are distributed post factum.
+	// I.e., first increase the filled epoch number, then distribute rewards during the epoch.
+	if !stream.DistributeTo.TotalWeight.IsZero() {
+		stream.FilledEpochs += 1
 	}
 
+	// TODO: can we delete this event?
```

Comment:
if we don't end up deleting it, please use `uevent.EmitTypedEvent` as suggested above